### PR TITLE
QueryPage/redesign

### PR DESCRIPTION
**Pull request by JaroPowerH**
**Your checklist for this pull request**
- [x] I've read the [contributing guideline](https://github.com/CERT-Polska/mquery/blob/master/CONTRIBUTING.md).
- [x] I've tested my changes by building and running mquery, and testing changed functionality (if applicable)
- [ ] I've added automated tests for my change (if applicable, optional)
- [ ] I've updated documentation to reflect my change (if applicable)

**What is the current behaviour?**
More or less the same.

**What is the new behaviour?**
Some cosmetic changes and bug fixed.
E.g. Cosmetic change:
The same progress bar used in `/query` and `/recent` pages.
On `/query` it has 3 (rather than 4) sections. Status section is properly centered.
E.g. Bug fix:
In Matches table (/query), in case of long file names, badges were not visible.

Changes:

- All /recent page logic is in `QueryPage` container (all state and methods from other query components were moved to `QueryPage`). Other components are SFC.

- Potentially reusable code was moved to separate components. E.g. `ActionCopyToClipboard`, `ActionDownload`, `DownloadDropdown`, `ErrorPage`, `LoadingPage`, `QueryProgressBar`

- Font Awesome 4 were replaced by version 5 icons.

- Props type checking (via prop-types) was added to every  SFC.

- Conventions applied:
All handlers (methods called from outside) have `handle` prefix, e.g. `handleSubmitQuery`
Binding is made in constructor.
Property names for handlers have `on` prefix, e.g. `onParseQuery={handleParseQuery}`
Property/attribute names for boolean values  have `is` previx,  e.g.  `isCollapsed={this.state.isCollapsed}`
`const` is used for const variables.
All axios calls are asynchronous.
Functional components use arrow (lambda) syntax.

- The most of source code was placed in subdirectories:

```
./query
./recent
./config
./status
./components 
```
In `./query` all component responsible for /query:
`QueryPage` (container) -> `QueryLayoutManager`
`QueryLayoutManager` -> `QueryField`, `QueryParseStatus`, `QueryResultsStatus`
`QueryField` -> `QueryNavigation`, `QueryMonaco`
`QueryNavigation` -> `QuerySubmitNav`, `QueryEditParseNav`, `QuerySearchNav`
`QueryMonaco` -> `yara-lang`
`QueryResultsStatus` -> `QueryMatches` -> `QueryMatchesItem`

In `./recent` all component responsible for /recent:
`RecentPage` (container) -> `SearchJobs` -> `SearchJobItem`

In `./config` all component responsible for /config:
`ConfigPage` -> `ConfigEntries`

In `./status` all component responsible for /status:
`StatusPage` -> `BackendStatus`, `DatabaseTopology`, `VersionStatus`

In `.component` potentially reusable components:
`ActionCopyToClipboard`, `ActionDownload`, `DownloadDropdown`, `ErrorBoundary`, `ErrorPage`, `LoadingPage`, `QueryProgressBar`, `QueryTimer`

### CODE REVIEW

File: src/mqueryfront/src/recent/SearchJobItem.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+const SearchJobItem = (props) => {
+    const { job, onRemove, onCancel } = props;
+    const { id, status, submitted, rule_name, priority, files_matched } = job;
+    const rule_author = props.job.rule_author
+        ? props.job.rule_author
+        : "(no author)";
+    const isFinished = isStatusFinished(status);
+
```
Comment (msm-code) [inline src/mqueryfront/src/recent/SearchJobItem.js:36]: unnecessary newline?

File: src/mqueryfront/src/components/ActionDownload.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    const { href, tooltipMessage, size } = props;
+
+    return (
+        <a
+            href={href}
+            data-toggle="tooltip"
+            title={tooltipMessage}
+            className="text-secondary"
+        >
+            <i>
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/components/ActionDownload.js:17]: Without <i> this icon is visible in every Matches item line. With <i> it appears only on current (with focus) line.

File: src/mqueryfront/src/components/QueryProgressBar.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        files_processed,
+        total_files,
+        files_in_progress,
+        files_errored,
+        files_matched,
+    } = job;
+
+    const isFinished = isStatusFinished(status);
+    const inProgeressPc = getPercentage(files_in_progress, total_files);
+    const erroredPct = getPercentage(files_errored, total_files);
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/components/QueryProgressBar.js:28]: You are right.

File: src/mqueryfront/src/query/QueryPage.js [hunk 4]
```diff
@@ -183,81 +182,69 @@ class QueryPage extends Component {
...
-        this.props.history.push("/", { editQuery: this.queryHash });
-    }
-
     render() {
-        const queryParse = (
-            <QueryParseStatus
+        const pagination = this.state.job
+            ? {
+                  activePage: this.state.activePage,
+                  itemsCountPerPage: PAGE_SIZE,
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/query/QueryPage.js:220]: The structure of `pagination` object mimics `<Pagination `props. See:
```
            <Pagination
                activePage={pagination.activePage}
                itemsCountPerPage={pagination.itemsCountPerPage}
                totalItemsCount={pagination.totalItemsCount}
                pageRangeDisplayed={pagination.pageRangeDisplayed}
                onChange={pagination.onChange}

```

File: src/mqueryfront/src/query/QueryParseStatus.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+        const badge =
+            is_private || is_global ? (
+                <span className="badge badge-info">
+                    {is_private ? "private" : "global"}
+                </span>
+            ) : null;
+
+        return (
+            <div key={rule_name} style={{ marginTop: 20 }}>
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/query/QueryParseStatus.js:20]: Original line:
`<div key={rule.rule_name} style={{ marginTop: "55px" }}>`
It woul be beter to use:
`            <div key={rule_name} className="mt-3">
`

File: src/mqueryfront/src/recent/SearchJobs.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    const backendJobRows = jobs.map((job) => (
+        <SearchJobItem
+            key={job.id}
+            job={job}
+            onRemove={() => onRemove(job.id)}
+            onCancel={() => onCancel(job.id)}
+        />
+    ));
+
+    // make table itemsPerPage size
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/recent/SearchJobs.js:31]: You are right.

File: src/mqueryfront/src/recent/SearchJobs.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    let i = backendJobRows.length;
+    while (i < pagination.itemsCountPerPage) {
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/recent/SearchJobs.js:33]: Mayby `while` w/o` i `iterator:
    // make table itemsCountPerPage size
    while (backendJobRows.length < pagination.itemsCountPerPage) {
        backendJobRows.push(<SearchJobItemEmpty key={backendJobRows.length} />);
    }

File: src/mqueryfront/src/components/DownloadDropdown.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import { FONTAWESOMESIZES } from "./bootstrapUtils";
+import PropTypes from "prop-types";
+
+const DownloadDropdown = (props) => {
+    const { itemList, size } = props;
+    const dropdownItems = itemList.map((item, index) => {
+        const iconName = item.icon;
+
+        const icon =
+            iconName === "archive"
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/components/DownloadDropdown.js:17]: Mayby when number of possibilities is greater. I would be external function.
It is possible to have internal function in stateless functional component but it is bad practice (the function will be copied in every component).

File: src/mqueryfront/src/query/QueryParseStatus.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                <div className="form-group">
+                    <label>
+                        <span className="mr-2 font-weight-bold">
+                            {rule_name}
+                        </span>
+                        {badge}
+                    </label>
+                    <div
+                        className="rounded text-monospace text-break"
+                        style={{ backgroundColor: "#e9ecef" }}
```
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/query/QueryParseStatus.js:30]: The jumbotron has suitable background. It requeres only smaller padding.
Final version:
`                    <div className="jumbotron text-monospace text-break p-2">
`

File: src/mqueryfront/src/recent/SearchJobs.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (JaroPowerH) [reply src/mqueryfront/src/recent/SearchJobs.js:33]: `    // make table itemsCountPerPage size
    while (backendJobRows.length < pagination.itemsCountPerPage) {
        backendJobRows.push(<SearchJobItemEmpty key={backendJobRows.length} />);
    }
`

**Comment by ITAYC0HEN** [comment]
Hey @JaroPowerH! 👋  Some conflicts were introduced in the meantime, and this blocks the PR from being merged. Can you please resolve these issues? :)

**Comment by msm-code** [comment]
should've rebase instead of merging. Don't want to redo all the work now, so let's assume it's good enough (after squashing it'll look the same anyway)

**Pull request merged**